File: app/matestack/bootstrap/components/carousel.rb
class Bootstrap::Components::Carousel < Matestack::Ui::VueJsComponent
  vue_js_component_name "matestack-ui-bootstrap-carousel" 
  
  optional :start, :controls, :indicators, :fade 
  optional :items, class: { as: :bs_class }
  # event trigger
  optional :cycle_on, :pause_on, :prev_on, :next_on

  def setup
    @component_config["carousel-id"] = carousel_id
  end
  
  def response
    div carousel_attributes do
      # carousel indicator
      indicator_partial if indicators.present?
      # carousel content
      div class: "carousel-inner" do
        yield_components
        carousel_partial if items
      end
      controls_partial if controls
    end
  end

  protected

  def indicator_partial
    ol class: "carousel-indicators" do
      items.size.times do |index|
        li data: { target: carousel_id, 'slide-to': index }, 
          class: "#{'active' if index == (start || 0) }"
      end
    end
  end

  def carousel_partial
    items.each_with_index do |item, index|
      div class: "carousel-item #{'active' if index == (start || 0) }", 
        data: { interval: item[:interval] } do
        img class: "d-block w-100", path: item[:path]
        if item[:title] || item[:text]
          div class: "carousel-caption d-none d-md-block" do
            heading size: 5, text: item[:title]
            paragraph text: item[:text]
          end
        end
      end
    end
  end

  def controls_partial
    link class: "carousel-control-prev", path: "##{carousel_id}", data: { slide: :prev }, attributes: { 'role': "button" } do
      span class: "carousel-control-prev-icon", attributes: { 'aria-hidden': "true" }
      span class: "sr-only", text: "Previous"
    end
    link class: "carousel-control-next", path: "##{carousel_id}", data: { slide: :next }, attributes: { 'role': "button" } do
      span class: "carousel-control-next-icon", attributes: { 'aria-hidden': "true" }
      span class: "sr-only", text: "Next"
    end
  end

  def carousel_attributes
    html_attributes.merge(
      id: carousel_id,
      class: carousel_classes,
      data: { ride: 'carousel' }
    )
  end

  def carousel_classes
    [].tap do |classes|
      classes << 'carousel slide'
      classes << 'carousel-fade' if fade
      #custom classes 
      classes << bs_class
    end.join(' ').strip
  end

  def carousel_id
    @carousel_id ||= "matestack-carousel-#{SecureRandom.hex}"
  end

end

File: app/matestack/bootstrap/components/carousel.js
const bootstrap = require('bootstrap')

MatestackUiCore.Vue.component('matestack-ui-bootstrap-carousel', {
  mixins: [MatestackUiCore.componentMixin],
  data() {
    return {
      carouselInstance: undefined
    };
  },

  methods: {
    cycle: function(){
      const self = this;
      self.carouselInstance.cycle()
    },
    pause: function(){
      const self = this;
      self.carouselInstance.pause()
    },
    prev: function(){
      const self = this;
      self.carouselInstance.prev()
    },
    next: function(){
      const self = this;
      self.carouselInstance.next()
    }
  },

  mounted: function() {
    const self = this;
    var myCarousel = self.$el;
    self.carouselInstance = new bootstrap.Carousel(myCarousel, {
      interval: (self.componentConfig['interval'] || 5000)
    })
  },

  created: function() {
    const self = this
    if(self.componentConfig["cycle_on"] != undefined){
      var cycle_events = self.componentConfig["cycle_on"].split(",")
      cycle_events.forEach(cycle_event => MatestackUiCore.matestackEventHub.$on(cycle_event.trim(), self.cycle));
    }
    if(self.componentConfig["pause_on"] != undefined){
      var pause_events = self.componentConfig["pause_on"].split(",")
      pause_events.forEach(pause_event => MatestackUiCore.matestackEventHub.$on(pause_event.trim(), self.pause));
    }
    if(self.componentConfig["prev_on"] != undefined){
      var prev_events = self.componentConfig["prev_on"].split(",")
      prev_events.forEach(prev_event => MatestackUiCore.matestackEventHub.$on(prev_event.trim(), self.prev));
    }
    if(self.componentConfig["next_on"] != undefined){
      var next_events = self.componentConfig["next_on"].split(",")
      next_events.forEach(next_event => MatestackUiCore.matestackEventHub.$on(next_event.trim(), self.next));
    }
  },

  beforeDestroy: function() {
    const self = this
    if(self.componentConfig["cycle_on"] != undefined){
      var cycle_events = self.componentConfig["cycle_on"].split(",")
      cycle_events.forEach(cycle_event => MatestackUiCore.matestackEventHub.$off(cycle_event.trim(), self.cycle));
    }

    if(self.componentConfig["pause_on"] != undefined){
      var pause_events = self.componentConfig["pause_on"].split(",")
      pause_events.forEach(pause_event => MatestackUiCore.matestackEventHub.$off(pause_event.trim(), self.pause));
    }

    if(self.componentConfig["prev_on"] != undefined){
      var prev_events = self.componentConfig["prev_on"].split(",")
      prev_events.forEach(prev_event => MatestackUiCore.matestackEventHub.$off(prev_event.trim(), self.prev));
    }
    if(self.componentConfig["next_on"] != undefined){
      var next_events = self.componentConfig["next_on"].split(",")
      next_events.forEach(next_event => MatestackUiCore.matestackEventHub.$off(next_event.trim(), self.next));
    }
  },
});
